### PR TITLE
Teuchos warnings

### PR DESCRIPTION
**Pull request by ibaned**
@trilinos/teuchos

## Description
Silence some warnings by separating code and replacing a functor with a lambda.

## Motivation and Context
These warnings pop up in CUDA builds of Trilinos.
This is all towards the lofty goal of compiling with `-Werror`.

## How Has This Been Tested?

A CUDA build of Trilinos I have emits no Teuchos warnings, and a non-CUDA build passes all Teuchos tests.

### CODE REVIEW

File: packages/teuchos/core/src/Teuchos_RCPNode.cpp [hunk 1]
```diff
@@ -396,7 +380,16 @@ void RCPNodeTracer::printActiveRCPNodes(std::ostream &out)
...
+      std::sort(rcp_node_vec.begin(), rcp_node_vec.end(),
+          [] (const rcp_node_list_t::value_type &v1, const rcp_node_list_t::value_type &v2)
+          {
+#ifdef TEUCHOS_DEBUG
+            return v1.second.nodePtr->insertion_number() < v2.second.nodePtr->insertion_number();
+#else
+            return v1.first < v2.first;
+#endif
+          }
+      );
```
Comment (bartlettroscoe) [inline packages/teuchos/core/src/Teuchos_RCPNode.cpp]: So it complains about a function option but not a lambda function?

Comment (bartlettroscoe) [reply packages/teuchos/core/src/Teuchos_RCPNode.cpp]: But I like how much more compact the code is.  I am going to love C++11 (now that I am actually allowed to start using it).

Comment (ibaned) [reply packages/teuchos/core/src/Teuchos_RCPNode.cpp]: Yea, I think this is a CUDA quirk: the CUDA compiler compiles the predicate class but not the actual code that uses it to sort later on. Thats my best guess, it definitely looks weird to me. Anyway, I agree, the lambda is a win either way.

**Comment by ibaned** [comment]
@allevin why did this test fail?

**Comment by mhoemmen** [comment]
Thanks @ibaned ! :D

**Pull request merged**